### DISCUSSION TITLE
Update installed code by downloading .bin file from web server via Ethernet connection on Nano RP2040 Connect with mbed ?

### DISCUSSION THREAD

**Discussion by CantStopMyBrain** [JAndrassy/ArduinoOTA#245] [Q&A] [answered]
By cobbling together bits from the examples for this library I thought I was going to pull this off: downloading an update .bin file from a web server to update a Nano RP2040 Connect running mbed and using an SPI-connected W5500 Ethernet shield for network connection. Should this be possible to do? Or are there obstacles/conflicts that make this particular combination -- NanoRP2040/mbed/SPI/ethernet/etc -- a non-starter?

I combined the download part of a download example with the "custom storage" example for the Nano RP2040 to get a file from a web server and save it to the on-board flash storage. Almost everything appears to be working: I get the .bin file and write it to flash without any errors, and then successfully restart/reboot the system, but it continues to run the existing sketch, not the new one it downloaded.

The download examples (and the ArduinoIoTCloud library which the CustomStorage example mentions) all perform some kind of .apply() and/or reset/reboot after downloading and saving the file. I've tried calling mbed_watchdog_trigger_reset(), NVIC_SystemReset(), and a reset method given here: https://gershnik.github.io/2024/02/12/rp2040-prog-reset.html -- all of these successfully reset/restart/reboot the system, but when it comes up it runs the same code it was already running, not the downloaded update.

I looked at the SFU library but its apply() method is empty (only a return statement), and the .begin() method looks like it clears/reformats the area where the download goes; I tried calling each of those (in separate trials and together) but got the same result -- the existing sketch restarted.

I feel like the part I'm missing is, having downloaded the new .bin file to flash, somehow causing it to be loaded into the working-program area where it will be executed on restart. Or perhaps I need to reconfigure things before restarting, like turning off the ethernet SPI or something? Or perhaps I'm not downloading the correct file? When I "Export Compiled Binary" I wind up with several files: .bin, .elf, .uf2, .elf.uf2, .hex, .map... I've been using the .bin version because the FileSystemStorage.h file in the CustomStorage example writes to "/ota/UPDATE.BIN", and the ArduinoIoTCloud library writes a compressed version of the .bin file. Is that the correct file to download for an update?

If it's not obvious already, all of this is pretty new terrain for me, so apologies if I'm expressing things unclearly or overlooking some obvious obstacle or solution. Thanks in advance for any assistance or clarification.

**Comment by CantStopMyBrain** [comment]
PS -- I *did* do this step mentioned in the README:  "platform.local.txt from extras folder has to be copied into boards package installation folder. On my Ubuntu system, that meant copying from here:
~/Arduino/libraries/ArduinoOTA/extras/rp2040/platform.local.txt 
to here:
~/.arduino15/packages/arduino/hardware/mbed_nano/4.1.1/
I was not 100% sure that was the correct destination folder, but there were similar files in that folder already, such as platform.txt

I also noted the instruction "For RP2040 the bundled ArduinoOTA library must be deleted" but I did not find any ArduinoOTA other than the one installed when I asked for it in the IDE.

**Comment by JAndrassy** [comment]
This ArduinoOTA library only works with the Pico Core (and the ArduinoOTA in Pico Core must be deleted).

The CustomStorage example is for the Mbed Core, which has the SFU library to boot load from file system in flash memory.
The CustomStorage example while in this library, doesn't actually use it.

**Reply by CantStopMyBrain** [reply]
Thanks for the prompt reply. Indeed the sketch I'm working on does NOT use ArduinoOTA, but it does use code from two of your examples: the download code from OTASketchDownload, and FileSystemStorage.h from CustomStorageNanoConnect. So I was hoping you had enough familiarity and knowledge that you could provide the "missing link" I seem to have: Do you know what step(s) is(are) needed to actually load the file I've saved to flash so that it replaces the current sketch and runs after restart? It would appear possible, judging from the ArduinoIoTCloud library but I seem to be missing some detail... As you say, I thought the SFU library would bootload the code from flash, but that's not happening... am I missing a call somewhere?

**Reply by JAndrassy** [reply, marked as answer]
does the CustomStorageNanoConnect example work for you?
do both sketches, the running and the update have the SFU?

**Reply by CantStopMyBrain** [reply]
> do both sketches, the running and the update have the SFU?

BINGO!
The original update .bin I was using was just a simple blink sketch that would be easy to recognize... no SFU and no joy.
When I changed the update .bin to be a variation on the download-and-install sketch, which includes SFU, it worked!
In hindsight I can see why this matters, but it simply never occurred to me that some of the important action was going to happen *after* the restart. Thank you sooo much for your help! I'm going to look for your tip jar now ... ;)

**Reply by JAndrassy** [reply]
SFU checks if the sketch is larger than the SFU bin itself. a Blink sketch is smaller so it ignored it.

**Reply by CantStopMyBrain** [reply]
Oh, that's important to know... and a little surprising. So if an update results in a smaller .bin file than its predecessor, it won't be installed? It doesn't seem obvious to me that every time one updates code it's guaranteed to get larger. I wonder why it doesn't check for "is different size" rather than "is larger"? 
Do you have a link to the SFU source? I'd be curious to compare with what I find in my installed SFU library. SFU.h/.cpp are mystifyingly brief -- they define empty apply() and download() methods, and a begin() method which just mounts and possibly reformats the flash block device. There's a folder of extras that look at a glance to be for a stand-alone app and don't seem to apply the size-comparison logic you describe.
I'm mostly happy to have a functioning solution, but I don't want to overlook potential pitfalls down the road so I'd like to gain a better understanding of SFU.
Thanks again!

**Reply by CantStopMyBrain** [reply]
Wait... can you clarify: is the presence of SFU in the update .bin essential or irrelevant? Was it just the increase in size which caused it to work, or the presence of SFU in the update, or both?

**Reply by JAndrassy** [reply]
SFU appends a second stage bootloader before the normal sketch in the binary. and that second stage bootloader has to cut itself from the update file. so it can't load a file smaller than the part which should be cut.
the source of that second stage bootloader is in the `extra` folder in the SFU library

**Reply by CantStopMyBrain** [reply]
The updates are working perfectly, thank you! I'd like to ask a follow-up question here that is related to the update process but not specifically the ArduinoOTA library:
Is there a way to persist data (e.g., a file written with your FileSystemStorage class) on the RP2040 flash so that it survives a sketch update? I did a trial and the data survived a reset but not an update. Is it just a matter of finding some unused corner of the flash? Or does all of the 16MB of flash get erased during an update? Any info/suggestions/links to guide me would be appreciated, or if you know it's not possible then that would be disappointing but important to know, too! Thank you!

**Reply by JAndrassy** [reply]
FileSystemStorage  only wraps Mbed Core file system functions. the code is taken from the Arduino Cloud OTA library.
I don't know how the Mbed Core file system in flash works.

**Reply by CantStopMyBrain** [reply]
Thanks for your response. Having asked the question here, and having found a solution, I might as well outline it here for closure:

I was unsuccessful (probably not smart enough! ;) trying to generalize (with a parameterized constructor) your FileSystemStorage library to work with caller-specified memory addresses, so I turned instead to an [mbed/RP2040 version of LittleFS](https://github.com/khoih-prog/LittleFS_Mbed_RP2040), configured with #define RP2040_FLASH_SIZE (15 * 1024 * 1024); this causes it to use the lower 15MB of the total RP2040 16MB flash memory for placing a (default) 64kB filesystem, avoiding the last (top) 1MB of flash that is used by the FileSystemStorage and SFU libraries for OTA updates. I was able to create a LittleFS file that persisted through power resets, through IDE-driven sketch updates, and through sketch-driven updates via binaries downloaded from a server.

It's nice to have gotten this working, as it gives us one way to persist device-specific information (most critically, per-device wifi credentials as these devices are scattered worldwide, each on their own wifi network) through code updates without having to maintain per-device binaries with that info baked in. However, I don't think it's an optimal solution: it relies on magic addresses and the file/info in flash is not visible/accessible from the outside, only to the RP2040. So we will be exploring adding a small SD card shield instead, which should give us all the same capabilities but in a more robust, familiar, and accessible manner, albeit with added cost (~$10 per device for the shield and an SD card) and layout complexity.

**Reply by mikeme2009** [reply]
Hi CantStopMyBrain. 
Could you describe in detail the steps that must be taken to make the firmware update work? I use an ESp32 connected to the Mega250 via Serial2 and each one runs a different program. Thanks

**Reply by JAndrassy** [reply]
@mikeme2009 this discussion is about the "Nano RP2040 Connect" board with Mbed Core, which is not supported by the library

**Reply by mikeme2009** [reply]
> @mikeme2009 this discussion is about the "Nano RP2040 Connect" board with Mbed Core, which is not supported by the library

Hi JAndrassy. I wanted to see if I can make certain changes to CantStopMyBrain's project so that it works for my configuration. At the moment I am downloading the .bin file to the esp32 and sending it to the serial2 but I don't know what I could do next to upload it to the Mega2560.

**Reply by JAndrassy** [reply]
Mega is supported by the library so the way to do an update is different. see the advanced examples

**Reply by mikeme2009** [reply]
> Mega is supported by the library so the way to do an update is different. see the advanced examples

Thank you. I'll check the examples and if I can't finish the project, can I bother you?

**Reply by CantStopMyBrain** [reply]
It sounds like you (mikeme2009) may need a different approach because of the different board, but if you think it would be helpful to have further details of the solution I worked out let me know what specifically you are looking for and I'll do my best to answer.
